filebeat/input/filestream/fswatch_test.go

```diff
@@ -19,301 +19,698 @@ package filestream
+		expEvent := loginp.FSEvent{
+			OldPath: filename,
+			Op:      loginp.OpDelete,
+			Descriptor: loginp.FileDescriptor{
+				Filename: filename,
+				Info:     testFileInfo{path: basename, size: 2},
 			},
-			nextFiles: map[string]os.FileInfo{
-				"path_new": testFileInfo{"path_new", 10, newTs, nil},
+		}
+		requireEqualEvents(t, expEvent, e)
+	})
+
+	t.Run("propagates a fingerprints for a new file", func(t *testing.T) {
+		dir := t.TempDir()
+		paths := []string{filepath.Join(dir, "*.log")}
+		cfgStr := `
+scanner:
+  check_interval: 10ms
+  symlinks: false
+  recursive_glob: true
+  fingerprint:
+    enabled: true
+    offset: 0
+    length: 1024
+`
+
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+
+		fw := createWatcherWithConfig(t, paths, cfgStr)
+		go fw.Run(ctx)
+
+		basename := "created.log"
+		filename := filepath.Join(dir, basename)
+		err := os.WriteFile(filename, []byte(strings.Repeat("a", 1024)), 0777)
+		require.NoError(t, err)
+
+		e := fw.Event()
+		expEvent := loginp.FSEvent{
+			NewPath: filename,
+			Op:      loginp.OpCreate,
+			Descriptor: loginp.FileDescriptor{
+				Filename:    filename,
+				Fingerprint: "2edc986847e209b4016e141a6dc8716d3207350f416969382d431539bf292e4a",
+				Info:        testFileInfo{path: basename, size: 1024},
 			},
-			expectedEvents: []loginp.FSEvent{
-				{Op: loginp.OpDelete, OldPath: "path_deleted", NewPath: "", Info: testFileInfo{"path_deleted", 5, oldTs, nil}},
-				{Op: loginp.OpCreate, OldPath: "", NewPath: "path_new", Info: testFileInfo{"path_new", 10, newTs, nil}},
+		}
+		requireEqualEvents(t, expEvent, e)
+	})
+
+	t.Run("does not emit events if a file is touched and resend_on_touch is disabled", func(t *testing.T) {
+		dir := t.TempDir()
+		paths := []string{filepath.Join(dir, "*.log")}
+		cfgStr := `
+scanner:
+  check_interval: 10ms
+`
+
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+		defer cancel()
+
+		fw := createWatcherWithConfig(t, paths, cfgStr)
+		go fw.Run(ctx)
+
+		basename := "created.log"
+		filename := filepath.Join(dir, basename)
+		err := os.WriteFile(filename, []byte(strings.Repeat("a", 1024)), 0777)
+		require.NoError(t, err)
+
+		e := fw.Event()
+		expEvent := loginp.FSEvent{
+			NewPath: filename,
+			Op:      loginp.OpCreate,
+			Descriptor: loginp.FileDescriptor{
+				Filename: filename,
+				Info:     testFileInfo{path: basename, size: 1024},
 			},
-		},
-	}
+		}
+		requireEqualEvents(t, expEvent, e)
 
-	for name, test := range testCases {
-		test := test
+		time := time.Now().Local().Add(time.Hour)
+		err = os.Chtimes(filename, time, time)
+		require.NoError(t, err)
 
-		t.Run(name, func(t *testing.T) {
-			w := fileWatcher{
-				log:          logp.L(),
-				prev:         test.prevFiles,
-				scanner:      &mockScanner{test.nextFiles},
-				events:       make(chan loginp.FSEvent),
-				sameFileFunc: testSameFile,
-			}
+		e = fw.Event()
+		require.Equal(t, loginp.OpDone, e.Op)
+	})
+}
```

Comment:
Would it be possible to add a test case here that creates a file, checks that the create event is returned for it, then creates another file in the same folder with the same fingerprint, and checks that no event is returned for it's creation?  Essentially the test case would show that the fingerprinting feature is working as expected, with duplicate data not being ingested.

Reply:
Done in https://github.com/elastic/beats/pull/35734/commits/d49fb7762d80916f244675dd69bd65a9e1623f8c